src/components/sbb-pearl-chain/sbb-pearl-chain.tsx

```diff
@@ -9,91 +10,141 @@ import legsData from './sbb-pearl-chain.helper';
-   * `cancellation`: if set, the leg will be marked as canceled.
    */
 
-  @Prop() public legs!: string;
+  @Prop() public legs!: InterfacePearlChainAttributes['legs'];
 
   /**
    * Per default, the current location has a pulsating animation. You can
    * disable the animation with this property.
    */
   @Prop() public disableAnimation?: boolean;
 
-  public render(): JSX.Element {
-    const legs = legsData(this.legs);
+  private _departureTime = this.legs && toDate(this.legs[0]?.departure?.time);
+  private _arrivalTime = this.legs && toDate(this.legs[this.legs?.length - 1].arrival?.time);
 
-    const statusClass = this.status === 'past' ? ' pearl-chain--past' : '';
+  private _getAllDuration(legs: InterfacePearlChainAttributes['legs']): number {
+    return legs.reduce((sum: number, leg) => (sum += leg.duration), 0);
+  }
 
-    let departureCancelClass = '';
-    let arrivalCancelClass = '';
+  private _getRelativeDuration(legs: InterfacePearlChainAttributes['legs'], leg: Leg): number {
+    const allDurations = this._getAllDuration(legs);
 
-    if (legs.length > 0) {
-      departureCancelClass = legs[0].cancellation ? ' pearl-chain--departure-cancellation' : '';
+    return Math.round((leg.duration / allDurations) * 100);
+  }
 
-      if (legs.length > 1) {
-        arrivalCancelClass = legs[legs.length - 1].cancellation
-          ? ' pearl-chain--arrival-cancellation'
-          : '';
-      }
+  private _getProgress(start: Date, end: Date): number {
+    const progress = differenceInMinutes(Date.now(), start);
+    const total = differenceInMinutes(end, start);
 
-      if (legs.length === 1) {
-        arrivalCancelClass = legs[0].cancellation ? ' pearl-chain--arrival-cancellation' : '';
-      }
+    return Math.round((progress / total) * 100);
+  }
+
+  private _getStatus(start: Date, end: Date): string {
+    if (isFuture(start)) {
+      return 'future';
+    }
+    if (isPast(start) && isPast(end)) {
+      return 'past';
     }
+    if (isPast(start) && isFuture(end)) {
+      return 'progress';
+    }
+  }
 
-    const classes = `pearl-chain${statusClass}${departureCancelClass}${arrivalCancelClass}`;
-    const statusIsRunning = this.status && this.status !== 'past' && this.status !== 'future';
+  private _renderPosition(): JSX.Element {
+    const currentPosition = this._getProgress(this._departureTime, this._arrivalTime);
 
-    if (statusIsRunning) {
-      if (this.status > 100) {
-        this.status = 100;
-      } else if (this.status < 0) {
-        this.status = 0;
+    const statusStyle = (): { [key: string]: string } => {
+      if (currentPosition >= 10) {
+        return {
+          '--status-position': `${currentPosition}%`,
+          transform: `translateX(-100%)`,
+        };
       }
+
+      return {
+        '--status-position': `${currentPosition}%`,
+      };
+    };
+
+    const animation = this.disableAnimation
+      ? this.disableAnimation === true
```

Comment:
Why is this nested?

const animation = this.disableAnimation ? 'pearl-chain__position--no-animation' : ''